### PR TITLE
Update KeyframeAnimationEffect to reflect FPWD

### PR DESCRIPTION
**Pull request by steveblock**
- Update KeyframeAnimationEffect API
- Remove Keyframe from API as this is now just a dictionary
- Update algorithm for normalizing the list of keyframes
- Update algorithm for sampling a KeyframeAnimationEffect
- Handle interpolating between keyframes with different composite operations
- Remove redundant KeyframeList and GroupedAnimationEffect types
- Remove merge composite operation
- Update tests accordingly

### CODE REVIEW

**Comment by steveblock** [comment]
This isn't quite ready for landing yet.

test-compositor.html needs fixing, but I'm waiting for the manual testing mode to be resurrected to help with debugging.

File: web-animations.js [hunk 1]
```diff
@@ -1168,177 +1086,270 @@ PathAnimationEffect.prototype = createObject(AnimationEffect.prototype, {
...
 /** @constructor */
-var KeyframeAnimationEffect =
-    function(property, operation, accumulateOperation) {
-  AnimationEffect.call(this, constructorToken, operation, accumulateOperation);
-  this.property = property;
-  this.frames = new KeyframeList(constructorToken);
+var KeyframeAnimationEffect = function(oneOrMoreFramesAsProperties, composite, accumulate) {
+  AnimationEffect.call(this, constructorToken, composite, accumulate);
+  if (!Array.isArray(oneOrMoreFramesAsProperties)) {
+    oneOrMoreFramesAsProperties = [ oneOrMoreFramesAsProperties ];
```
Comment (dstoc) [inline web-animations.js]: Don need the spaces here, just [foo] (see http://google-styleguide.googlecode.com/svn/trunk/javascriptguide.xml?showone=Code_formatting#Code_formatting)

File: web-animations.js [hunk 2]
```diff
@@ -1168,177 +1086,270 @@ PathAnimationEffect.prototype = createObject(AnimationEffect.prototype, {
...
+  }
+  this.frames = [];
```
Comment (dstoc) [inline web-animations.js]: this.frames = oneOrMoreFramesAsProperties.map(Keyframe.createFromProperties); ?

File: web-animations.js [hunk 3]
```diff
@@ -1168,177 +1086,270 @@ PathAnimationEffect.prototype = createObject(AnimationEffect.prototype, {
...
-        return;
+    var properties = this._getProperties();
+    for (var i = 0; i < properties.length; i++) {
+      // TODO: Handle accumulation.
+      compositor.setAnimatedValue(target, properties[i],
+          this._sampleForProperty(timeFraction, properties[i]));
+    }
+  },
+  _sampleForProperty: function(timeFraction, property) {
+    // Get the property-specific keyframes.
```
Comment (dstoc) [inline web-animations.js]: Get -> Find ?

File: web-animations.js [hunk 4]
```diff
@@ -1168,177 +1086,270 @@ PathAnimationEffect.prototype = createObject(AnimationEffect.prototype, {
...
-  },
   toString: function() {
     return this.property;
+  },
+  _compositeForKeyframe: function(keyframe) {
+    return isDefinedAndNotNull(keyframe.composite) ?
+        keyframe.composite : this.composite;
+  },
+  _areKeyframesLooselySorted: function() {
+    var maximumOffset = -Math.Infinity;
```
Comment (dstoc) [inline web-animations.js]: Infinity doesn't seem to be on Math. Just '-Infinity' ?

Comment (dstoc) [inline web-animations.js]: This was slightly confusing. I would have called this previousOffset.

File: web-animations.js [hunk 5]
```diff
@@ -1168,177 +1086,270 @@ PathAnimationEffect.prototype = createObject(AnimationEffect.prototype, {
...
+  if (!Array.isArray(oneOrMoreFramesAsProperties)) {
+    oneOrMoreFramesAsProperties = [ oneOrMoreFramesAsProperties ];
+  }
+  this.frames = [];
+  for (var i = 0; i < oneOrMoreFramesAsProperties.length; i++) {
+    this.frames.push(Keyframe.createFromProperties(oneOrMoreFramesAsProperties[i]));
+  }
+  // Set lazily
+  this._properties = null;
+  this._normalizedFrames = null;
```
Comment (dstoc) [inline web-animations.js]: Add "cached" to the name? I assume this will be invalidated when we call setFrames

File: web-animations.js [hunk 6]
```diff
@@ -2650,16 +2660,80 @@ var toCssValue = function(property, value, svgMode) {
...
 
 var fromCssValue = function(property, value) {
+  if (value === neutralCssValue) {
+    return neutralValue;
+  }
   return getType(property).fromCssValue(value);
 }
 
+// Sentinel values
+var neutralCssValue = null;
```
Comment (dstoc) [inline web-animations.js]: Should this be unique? neutralCssValue = {};

File: web-animations.js [hunk 7]
```diff
@@ -2650,16 +2660,80 @@ var toCssValue = function(property, value, svgMode) {
...
+        'Derived classes must override CompositableValue.compositeOnto()');
+  },
+  // This is purely an optimization.
+  dependsOnUnderlyingValue: function() {
+    return true;
+  },
+};
+
+
+var AddReplaceCompositableValue = function(value, composite) {
```
Comment (dstoc) [inline web-animations.js:2868]: @constructor ?

File: web-animations.js [hunk 8]
```diff
@@ -2650,16 +2660,80 @@ var toCssValue = function(property, value, svgMode) {
...
+  compositeOnto: function(property, underlyingValue) {
+    switch (this.composite) {
+      case 'replace':
+        // TODO: Is it correct that replace-compositing the neutral value yields
+        // the neutral value?
+        return this.value;
+      case 'add':
+        return add(property, underlyingValue, this.value);
+      default:
+        throw new Error('Logic error: Invalid composite operation ' +
```
Comment (dstoc) [inline web-animations.js]: Perhaps we should define an assertNotReached function.

File: web-animations.js (hunk 6 above)
Comment (steveblock) [inline web-animations.js]: Not sure what you mean by this?

**Comment by steveblock** [comment]
Have now fixed test-compositor.html

Ready for another look

File: web-animations.js [hunk 9]
```diff
@@ -1118,13 +1118,13 @@ MediaReference.prototype = createObject(TimedItem.prototype, {
...
-var AnimationEffect = function(token, operation, accumulateOperation) {
+var AnimationEffect = function(token, composite, accumulate) {
   if (token !== constructorToken) {
     throw new TypeError('Illegal constructor');
   }
-  this.operation = operation === undefined ? 'replace' : operation;
-  this.accumulateOperation =
-      accumulateOperation == undefined ? 'none' : accumulateOperation;
+  // Use the default value if an invalid string is specified.
+  this.composite = composite === 'add' ? 'add' : 'replace';
```
Comment (dstoc) [inline web-animations.js:1126]: Perhaps an issue in the spec, but it appears that composite is only defined on KeyframeAnimationEffect.

File: web-animations.js [hunk 10]
```diff
@@ -1307,178 +1227,310 @@ PathAnimationEffect.prototype = createObject(AnimationEffect.prototype, {
...
-  AnimationEffect.call(this, constructorToken, operation, accumulateOperation);
-  this.property = property;
-  this.frames = new KeyframeList(constructorToken);
+var KeyframeAnimationEffect = function(oneOrMoreKeyframesDictionaries,
+    composite, accumulate) {
+  AnimationEffect.call(this, constructorToken, composite, accumulate);
+  if (!Array.isArray(oneOrMoreKeyframesDictionaries)) {
+    oneOrMoreKeyframesDictionaries = [oneOrMoreKeyframesDictionaries];
+  }
+  this.keyframeDictionaries =
```
Comment (dstoc) [inline web-animations.js]: _keyframeDictionaries

File: web-animations.js [hunk 11]
```diff
@@ -1307,178 +1227,310 @@ PathAnimationEffect.prototype = createObject(AnimationEffect.prototype, {
...
+
+    if (frames.length === 0) {
+      return new AddReplaceCompositableValue(rawNeutralValue, 'add');
+    }
+
+    // Add 0 and 1 keyframes if required.
+    if (frames[0].offset !== 0.0) {
+      var keyframe = new KeyframeInternal(0.0, 'add');
+      keyframe.addPropertyValuePair(property, cssNeutralValue);
+      frames.splice(0, 0, keyframe);
```
Comment (dstoc) [inline web-animations.js]: frames.unshift ?

File: web-animations.js [hunk 12]
```diff
@@ -1307,178 +1227,310 @@ PathAnimationEffect.prototype = createObject(AnimationEffect.prototype, {
...
-    return result;
+    return this._cachedProperties;
+  }
+});
+
+
+// An internal representation of a keyframe. The Keyframe type from the spec is
+// just a dictionary and is not exposed.
+/** @constructor */
+var KeyframeInternal = function(offset, composite) {
```
Comment (dstoc) [inline web-animations.js:1492]: This is OK, but I think we have other internal types defined that are not marked *Internal.

File: web-animations.js [hunk 13]
```diff
@@ -1307,178 +1227,310 @@ PathAnimationEffect.prototype = createObject(AnimationEffect.prototype, {
...
+  console.assert(typeof offset === 'number' || offset === null,
```
Comment (dstoc) [inline web-animations.js:1493]: Can an API user trigger these to fail?

**Comment by dstoc** [comment]
lgtm

**Comment by mithro** [comment]
PS You need to rebase/merge due to other merges.

File: web-animations.js (hunk 10 above)
Comment (steveblock) [inline web-animations.js]: Done

File: web-animations.js (hunk 11 above)
Comment (steveblock) [inline web-animations.js]: Done

File: web-animations.js (hunk 13 above)
Comment (steveblock) [inline web-animations.js:1493]: No, they're just checks on internal logic.

File: web-animations.js (hunk 9 above)
Comment (steveblock) [inline web-animations.js:1126]: Sent out a spec patch - https://github.com/web-animations/web-animations-spec/pull/13

File: web-animations.js [hunk 14]
```diff
@@ -2758,11 +2810,10 @@ var getType = function(property) {
   return propertyTypes[property] || nonNumericType;
 }
 
-var zero = function(property, value) {
-  return getType(property).zero(value);
-};
-
```
Comment (shans) [inline web-animations.js:2764]: Do we use type-specific zeros anywhere any more? Maybe remove them from the type objects too?

Ah, I assume they're still called from interpolate?

File: web-animations.js [hunk 15]
```diff
@@ -2790,16 +2841,77 @@ var toCssValue = function(property, value, svgMode) {
...
+  this.composite = composite;
+};
+
+AddReplaceCompositableValue.prototype =
+    createObject(CompositableValue.prototype, {
+  compositeOnto: function(property, underlyingValue) {
+    switch (this.composite) {
+      case 'replace':
+        // TODO: Is it correct that replace-compositing the neutral value yields
+        // the neutral value?
```
Comment (shans) [inline web-animations.js:2879]: Yes.

File: web-animations.js (hunk 14 above)
Comment (steveblock) [inline web-animations.js:2764]: Some can be removed - https://github.com/web-animations/web-animations-js/pull/295

File: web-animations.js (hunk 15 above)
Comment (steveblock) [inline web-animations.js:2879]: Removed TODO in https://github.com/web-animations/web-animations-js/pull/295

**Pull request merged**